crates/autopilot/src/domain/competition/participation_guard/db.rs

```diff
@@ -0,0 +1,122 @@
+use {
+    crate::{
+        domain::{eth, Metrics},
+        infra,
+    },
+    ethrpc::block_stream::CurrentBlockWatcher,
+    std::{
+        collections::HashMap,
+        sync::Arc,
+        time::{Duration, Instant},
+    },
+};
+
+/// Checks the DB by searching for solvers that won N last consecutive auctions
+/// but never settled any of them.
+#[derive(Clone)]
+pub(super) struct Validator(Arc<Inner>);
+
+struct Inner {
+    persistence: infra::Persistence,
+    banned_solvers: dashmap::DashMap<eth::Address, Instant>,
+    ttl: Duration,
+    last_auctions_count: u32,
+    drivers_by_address: HashMap<eth::Address, Arc<infra::Driver>>,
+}
+
+impl Validator {
+    pub fn new(
+        persistence: infra::Persistence,
+        current_block: CurrentBlockWatcher,
+        competition_updates_receiver: tokio::sync::mpsc::UnboundedReceiver<()>,
+        ttl: Duration,
+        last_auctions_count: u32,
+        drivers_by_address: HashMap<eth::Address, Arc<infra::Driver>>,
+    ) -> Self {
+        let self_ = Self(Arc::new(Inner {
+            persistence,
+            banned_solvers: Default::default(),
+            ttl,
+            last_auctions_count,
+            drivers_by_address,
+        }));
+
+        self_.start_maintenance(competition_updates_receiver, current_block);
+
+        self_
+    }
+
+    /// Update the internal cache only once the competition auctions table is
+    /// updated to avoid redundant DB queries on each block or any other
+    /// timeout.
+    fn start_maintenance(
+        &self,
+        mut competition_updates_receiver: tokio::sync::mpsc::UnboundedReceiver<()>,
+        current_block: CurrentBlockWatcher,
+    ) {
+        let self_ = self.clone();
+        tokio::spawn(async move {
+            while competition_updates_receiver.recv().await.is_some() {
```

Comment:
This should always be `Some`, right? In that case I would put a `tracing::error!()` after the loop alerting that the loop exited.